DailyKit/CallManager/Models/CallParticipants.swift

```diff
@@ -0,0 +1,133 @@
+    }
+
+    struct Builder: Equatable {
+        private(set) var local: CallParticipant
+        private let limit: Int
+
+        init(local: CallParticipant, limit: Int = Constants.visibleParticipantLimit) {
+            self.local = local
+            self.limit = limit
+        }
+
+        private(set) var remote: [CallParticipant.ID: CallParticipant] = [:]
+        private(set) var speakerOrder: [CallParticipant.ID] = []
+        private(set) var joinOrder: [CallParticipant.ID] = []
+
+        var activeSpeaker: CallParticipant? {
+            didSet {
+                if let activeSpeaker, activeSpeaker.isLocal == false {
+                    speakerOrder.removeAll(where: { $0 == activeSpeaker.id })
+                    speakerOrder.append(activeSpeaker.id)
+                    speakerOrder = Array(speakerOrder.prefix(limit))
+                }
+            }
+        }
+
+        func build() -> CallParticipants {
+            var ids: Set<CallParticipant.ID> = []
+            var newVisible: [Int: CallParticipant] = [:]
+
+            func append(_ participant: CallParticipant) {
+                guard ids.count < limit else { return }
+                guard ids.contains(participant.id) == false else { return }
+
+                ids.insert(participant.id)
+                newVisible[newVisible.count] = participant
+            }
+
+            func append(_ participantIDs: [CallParticipant.ID]) {
+                var withVideo: [CallParticipant] = []
+                var withoutVideo: [CallParticipant] = []
+
+                let participants = participantIDs
+                    .map { remote[$0] }
+                    .compactMap { $0 }
+
+                for participant in participants {
+                    if participant.hasVideo {
+                        withVideo.append(participant)
+                    } else {
+                        withoutVideo.append(participant)
+                    }
+                }
+
+                withVideo.forEach { append($0) }
+                withoutVideo.forEach { append($0) }
+            }
+
+            if let activeSpeaker, activeSpeaker.isLocal == false {
+                append(activeSpeaker)
+            }
+
+            append(speakerOrder.reversed())
+            append(joinOrder.reversed())
+
+            return CallParticipants(
+                local: local,
+                remote: remote,
+                visible: newVisible
+            )
+        }
+
+        mutating func handleJoined(_ participant: CallParticipant) {
+            assert(participant.isLocal == false, "Unexpected join event for the local participant.")
+
+            joinOrder.append(participant.id)
+            joinOrder = Array(joinOrder.prefix(limit))
+
+            remote[participant.id] = participant
+        }
+
+        mutating func handleUpdated(_ participant: CallParticipant) {
+            if participant.isLocal {
+                local = participant
```

Comment:
Wouldn't this be up to date from the `callClient` reference it was originally initialized with? Or is that a copy?

Reply:
Participants are value types, so updates to a participant's fields do not propagate between copies.